### PR TITLE
Kernel: Move `PROCS` array to `Kernel` struct

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

As a part of #1043, this moves the PROCS array to no longer be a static global variable but instead be a member of the `Kernel` struct. This is needed for at least three reasons:

1. The `Process` type needs to be templated across architectures, and that doesn't seem possible with the static array (because the kernel crate doesn't know what type it will actually be).
2. This will help with the MPU changes.
3. It is generally better to not have global variables that need `unsafe` to access.

After #1109, this is relatively straightforward, with a couple things of note:

- To access the `Kernel` struct, many data structures need to keep a reference to the kernel struct, and these have been added.
- debug.rs cannot directly access the `PROCS` array, so it must be passed in during a panic.

~~This is blocked on #1046~~
### Testing Strategy

Running hail on hail.

### TODO or Help Wanted

  - [ ] Update CHANGELOG

  - [ ] Test

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
Ok! Please have a look at this PR.

**Comment by bradjc** [comment]
One idea that has been floated is changing how grants are created for capsules from:

```rust
let grant = kernel::Grant::create(board_kernel);
```

to 

```rust
let grant = board_kernel.create_grant();
```

That would have three benefits:

1. It makes the API a little more consistent, and makes it clear that `board_kernel` (the `&kernel::Kernel`) is the main access to the kernel API. `.kernel_loop` is also called from the `Kernel` struct.
2. It simplifies where the kernel API is defined. What operations does the kernel provide? Look at the `pub` functions in the `impl Kernel` block. Now this is different than the interfaces and tools the kernel provides (like `StaticRef`, `Component`, etc., and the HIL interfaces).
3. It would be straightforward to move the static variable `CONTAINER_COUNTER` to the `Kernel` struct. Doing this would allow the kernel to verify that no grants are created after processes are created (which would be invalid if any code tried to do that because no space would be reserved for them in the process).

Thoughts?

**Comment by alevy** [comment]
@bradjc I'm convinced

**Comment by bradjc** [comment]
If you want to see it in code form it is here: https://github.com/tock/tock/pull/1111/commits/70f53cafa2f51b3051e21051392b3d19695a508a

**Comment by bradjc** [comment]
Doc updates are in #1113, this pr is really more of a refactor.

**Comment by alevy** [comment]
This last change, from holding a reference to the `Kernel` to holding a direct reference to a `Process` is very problematic:

As is, there is nothing checking that this `Process` is indeed still the same one we thought it was when we created the `AppId`. It could have been replaced with a new instance of the same application (after a restart) or a different application entirely. At least going through the kernel gives us the flexibility to implement this (since an internal `idx` can just be an opaque identifier).

In general, it could be possible to just explicitly check that the process being referred to is the same version: replacing the idx field with a random/monotonically increasing value for each new instance of a process and comparing that with a field in the process struct.

However, indirecting through a `Kernel` just gives us way more flexibility about how to store processes

File: kernel/src/callback.rs [hunk 1]
```diff
@@ -1,30 +1,52 @@
...
 use core::ptr::NonNull;
-use process;
+
+use process::{self, Process};
 
 /// Userspace app identifier.
-#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+#[derive(Clone, Copy)]
 pub struct AppId {
+    crate process: &'static Process<'static>,
```
Comment (alevy) [inline kernel/src/callback.rs]: I think this is a dangerous change, and that it should be reverted to storing a reference to the `Kernel`

**Comment by bradjc** [comment]
Removed the commit.

It will probably make sense to move to a more opaque "idx" in AppId/Process soon.

**Comment by bradjc** [comment]
I added updating the changelog as a todo item in the tracking PR. Otherwise I think it will get rewritten multiple times with the series of PRs and any potential changes.

**Comment by dverhaert** [comment]
Looks good!

**Pull request merged**